Backend: Move page file and error counts to crawl replay.json endpoint

tw4l opened the pull request:
Backend work for https://github.com/webrecorder/browsertrix/issues/1859

- Remove file count from qa stats endpoint
- Add file and error counts to crawl replay.json endpoint
- Calculate counts on the fly once and then cache in database if crawl is successfully finished and all pages added

## Testing instructions

1. Run a crawl with files and unresolvable URLs (e.g. URL list workflow with `https://webrecorder.net`, `https://www.bitarchivist.net/static/usersurvey-1-demographics-9121af3ce1bce7c80e5643dbbc7b9f5e-29b25.png`, and `https://example.invalid` as the URLs)
2. When crawl completes, open dev tools, go to Replay tab in the Browsertrix UI, and check the `replay.json` response in the dev tools network tab. You should see as part of the response:

```
...
filePageCount: 1,
errorPageCount: 1,
...
```

tw4l committed:
Move page file and error counts to replay.json endpoint

- Remove file count from qa stats endpoint
- Add file and error counts to replay.json
- Calculate counts on the fly once and then cache in database
if crawl is successfully finished and all pages added

ikreymer commented on backend/btrixcloud/pages.py:
```diff
@@ -501,7 +501,11 @@ async def list_pages(
-    async def get_crawl_file_count(self, crawl_id: str):
+    async def get_crawl_page_count(self, crawl_id: str) -> int:
+        """Return total number of pages in db for crawl"""
+        return await self.pages.count_documents({"crawl_id": crawl_id})
```
This works, but should also already be available via the `CrawlStats.done` field in the crawl model, which should be equal to this. Any discrepancy (hopefully there shouldn't be) should be counted as an error, since this is the number that we display to users of how many pages were crawled / and total pages for QA run.

ikreymer replied:
Ah never mind, I see that this is being compared against to determine when all pages were added!

tw4l replied:
Yup! Just wanted to make sure we didn't cache incomplete numbers in the database :)

ikreymer commented:
Nice! This works for now, though another option would be to compute this incrementally when a page is being added, and then do a migration for existing crawls. I suppose that's something we can switch to at a later time, as well, though.

tw4l commented:
Per Ilya's note, converting this to add to counts as pages are added, with migration for older crawls.

tw4l committed:
Calculate crawl error and file counts while adding pages

tw4l committed:
Add and set Page.isFile and Page.isError fields

tw4l committed:
Add migration

tw4l commented:
Will convert back from draft when new approach and migration are tested and ready for re-review.

tw4l committed:
Fixups

tw4l committed:
Make Page.isFile and isError optional

tw4l committed:
Fix type warning

tw4l committed:
Add pylint disable

tw4l committed:
Fix updating crawl

tw4l committed:
Add Page.isFile and Page.isError to tests

tw4l commented:
Ready for re-review @ikreymer , I've tested the migration locally but we should test it on dev as well.

ikreymer committed:
simplify:
- add Page.compute_page_type() to compute isError / isFile in one place
- set isFile / isError before adding page to db
- reuse compute_page_type() in migration
- always use '$inc' to update pageFileCount and pageErrorCount, doesn't exist so will increment from 0

ikreymer commented:
Added some optimizations/cleanup, I think should be good to go!

ikreymer committed:
page type: also treat pages loaded to loadState == 2 and no title / status 200 and no mime as files

ikreymer committed:
tweak status check

ikreymer committed:
fix page type check again

ikreymer committed:
just check status == 200
